Refactor low power/manual power-down actor disabling

reaperrr opened the pull request:
This aims to remove the biggest remaining roadblock towards getting rid of Actor.Disabled and the IDisable interface.

The idea is to no longer disable the actor as such, but only the traits that need to be disabled, via a condition (or multiple conditions) granted either at specific power levels (usually anything below Normal) or manual power down.

reaperrr committed:
Refactor CanPowerDown to provide a condition instead of triggering IsDisabled

reaperrr committed:
Refactor PowerManager and RequiresPower to use conditions

Instead of Actor.IsDisabled.
Added INotifyPowerLevelChanged interface to do so as efficiently as possible.

reaperrr committed:
Adapt TS to power-related trait refactorings

reaperrr commented:
Note: Currently not working as intended (despite some first fixes suggested by @pchote), as affected actors are considered low power/disabled all the time, probably a bug somewhere in the `PowerManager` changes.

Only TS mod adapted on the yaml side so far, I first want to get the code debugged and the basic yaml approach approved before I adapt the other mods (this will also need an upgrade rule).

pchote commented on OpenRA.Mods.Common/Effects/PowerdownIndicator.cs:
```diff
@@ -33,7 +33,7 @@ public PowerdownIndicator(Actor a)
 		void IEffect.Tick(World world)
 		{
-			if (!a.IsInWorld || a.IsDead || !canPowerDown.Disabled)
+			if (!a.IsInWorld || a.IsDead || !canPowerDown.PoweredDown)
```
This whole class can be removed, and replaced by a conditional `WithDecoration` (see e.g. D2K carryall).

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnPowerState.cs:
```diff
@@ -0,0 +1,77 @@
+			conditionManager = self.TraitOrDefault<ConditionManager>();
+
+			// Set initial disabled state
+			Update(self);
```
This need to be called from `TraitEnabled`.  All traits start in the disabled state until some point after `Created` is called.  The `ConditionManager` will call `TraitEnabled` when it is ready (and if not starting disabled)

pchote commented on OpenRA.Mods.Common/Traits/Power/CanPowerDown.cs:
```diff
@@ -34,18 +39,47 @@ public class CanPowerDownInfo : ConditionalTraitInfo, Requires<PowerInfo>
+			conditionManager = self.TraitOrDefault<ConditionManager>();
+
+			// Set initial disabled state
+			Update(self);
```
Likewise here.

pchote commented on mods/ts/rules/gdi-structures.yaml:
```diff
@@ -322,11 +322,13 @@ GARADR:
 		IndicatorPalette: mouse
 		PowerupSpeech: EnablePower
 		PowerdownSpeech: DisablePower
+		PowerDownCondition: powered-down
```
`PowerDown` vs `Powerdown` vs `powered-down`.

reaperrr commented:
Updated, but the lint errors suggest there's still something wrong with how conditions are granted, maybe the error actually lies in the Update methods. I haven't found the cause yet, though.

pchote commented:
The lint errors look correct: actors that inherit `^DisabledOverlay` but not `^CanPowerDown` will consume `lowpower` and `powerdown` which are not granted.

reaperrr commented:
Rebased and updated.

The TS yaml setup is fixed now, the only remaining issue is that buildings don't recover from disabled state when power becomes positive again, so either `GrantConditionOnPowerState` or `PowerManager` doesn't notify/update as intended.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnPowerState.cs:
```diff
@@ -0,0 +1,84 @@
+
+		void Update(Actor self)
+		{
+			if (conditionManager == null || IsTraitDisabled)
```
Returning on `IsTraitDisabled` means that the `TraitDisabled` handing won't work.
Instead, I think you want to set `validPowerState = !IsTraitDisabled && Info.ValidPowerStates.HasFlag(playerPower.PowerState);`

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnPowerState.cs:
```diff
@@ -0,0 +1,84 @@
+			validPowerState = Info.ValidPowerStates.HasFlag(playerPower.PowerState);
+
+			if (validPowerState && conditionToken == ConditionManager.InvalidConditionToken)
+				conditionToken = conditionManager.GrantCondition(self, Info.Condition);
```
If `Update` is called and `validPowerState` has the same value that it did previously, then you will double grant or double revoke conditions.  This may be why buildings aren't returning from low power.

You should cache the last `validPowerState` and then exit early if it has not changed.

reaperrr commented on OpenRA.Game/Player.cs:
```diff
@@ -24,7 +24,14 @@
+	[Flags]
+	public enum PowerState
+	{
+		Normal = 0,
```
This appears to have been the problem, if I change them to 1, 2, 4 respectively, all problems are gone.

reaperrr committed:
PowerDown/low power refactor upgrade rule

reaperrr committed:
Adapt D2k to low power/power down refactor

reaperrr committed:
Make WithSpriteBody a pausable trait

Allowing to drop the PauseAnimationWhenDisabled property (in favor of using PausOnCondition).

reaperrr committed:
Adapt TD to low power/power-down refactor

Removed CanPowerDown from TD buildings.

Manual power-down is intentionally not supported by the in-game UI, the remaining CanPowerDown entries were effectively bit-rot.

reaperrr committed:
Adapt RA to low power/power-down refactor

reaperrr commented:
Fixed the bug and adapted all mods.

This is ready for fully-fledged reviews now.

pchote commented on mods/ra/rules/structures.yaml:
```diff
@@ -56,17 +58,27 @@ MSLO:
 	MustBeDestroyed:
 		RequiredForShortGame: false
 	WithNukeLaunchAnimation:
+	WithDecoration@POWERDOWN:
```
These should override the `ReferencePoint` to center the decoration:

![screen shot 2017-09-15 at 20 00 31](https://user-images.githubusercontent.com/167819/30499020-cddc93de-9a50-11e7-9613-8980a465fbc0.png)

I actually prefer it not zooming, but this does make it inconsistent with the repair glyph.  If we keep this behaviour then we'll probably need to block on porting that to `WithDecoration` too.

reaperrr replied:
> plus some kind of cycling condition trait so that only one decoration is shown at once

I don't agree with this part. Apart from sounding a bit too complicated for this, I highly doubt switching back and forth between the two would look good with the available art.

In my opinion we should just tweak the sequence offset(s) so they don't overlap, that way we don't need to scope-creep, don't need to deal with getting the transition right, and don't need to deal with players that would prefer to have both indicators visible all the time.

pchote replied:
Ok, that works too.

pchote replied:
The blinking can also be implemented at the `WithDecoration` trait level: `Created` can query and cache any other traits that declare the same position, and then `Tick` can cycle the visibility between these in turn.  If you want I could do a prototype demonstrating this.

reaperrr replied:
Fixed the reference points and disabled zooming on the RepairIndicator.

pchote commented on OpenRA.Mods.Common/Traits/Power/CanPowerDown.cs:
```diff
@@ -34,18 +34,43 @@ public class CanPowerDownInfo : ConditionalTraitInfo, Requires<PowerInfo>
+	public class CanPowerDown : ConditionalTrait<CanPowerDownInfo>, IPowerModifier, IResolveOrder, INotifyOwnerChanged
 	{
-		[Sync] bool disabled = false;
+		bool disabled = false;
```
We probably want to keep this `[Sync]`

pchote replied:
Can we please rename this to `poweredDown` to avoid confusion with `IsTraitDisabled`?

pchote commented on OpenRA.Mods.Common/Traits/Power/Player/PowerManager.cs:
```diff
@@ -107,15 +109,25 @@ public void Tick(Actor self)
+				wasLowPower = lowPower;
+				UpdateRequiresPowerActors();
+			}
+			else if (!lowPower && wasLowPower)
```
IMO something like the following would be cleaner:
```csharp
if (lowPower && !wasLowPower)
	nextPowerAdviceTime = 0;

wasLowPower = lowPower;

if (lowPower != wasLowPower)
	UpdateRequiresPowerActors();
```

pchote commented on mods/cnc/rules/structures.yaml:
```diff
@@ -557,6 +561,11 @@ HQ:
 	SupportPowerChargeBar:
 	Power:
 		Amount: -50
+	WithDecoration@POWERDOWN:
```
cnc doesn't support powering down, so these can all be removed.

pchote reviewed (commented):
Looking good overall, just a couple of nits on my first pass.  Also needs a rebase unfortunately.

reaperrr committed:
Don't scale RepairIndicator with zoom

reaperrr commented:
Rebased and updated.

Phrohdoh commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -1003,6 +1003,64 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
+						requiresPower.Value.Nodes.Add(new MiniYamlNode("Condition", "lowpower"));
+
+						Console.WriteLine("RequiresPower has been replaced with GrantConditionOnPowerState.");
+						Console.WriteLine("This new trait provides a condition if the current player power is valid.");
```
"... is valid" doesn't make sense to me here.

It toggles a condition based on power state. Saying anything other than that is confusing IMO.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnPowerState.cs:
```diff
@@ -0,0 +1,84 @@
+		[Desc("Condition to grant.")]
+		public readonly string Condition = null;
+
+		public readonly PowerState ValidPowerStates = PowerState.Low | PowerState.Critical;
```
Can we please make this `[FieldLoader.Require]` too?  Leaving this implicit makes the yaml side less clear than it should be IMO.

pchote commented on mods/d2k/rules/defaults.yaml:
```diff
@@ -474,5 +474,6 @@
 		RequiresCondition: disabled
 		Palette: disabled
-	GrantConditionOnDisabled@IDISABLE:
+	GrantCondition@IDISABLE:
```
Thinking out loud: Might be clearer to have a pair of templates for `^DisableOnLowPower` and `^DisableOnLowPowerOrPowerdown` instead of overriding `GrantCondition@IDISABLE` on powerdown-able actors?  These could then also include `GrantConditionOnPowerState`, `CanPowerDown`, and `WithDecoration@POWERDOWN`.

pchote reviewed (commented):
I also noticed the following issues with EMP in TS:
* EMPed radar doesn't disable radar
* EMPed laser fence nodes don't disable their connected segments

There may be other regressions in the other mods, but i'm sure they will be found and fixed in due time, so we probably shouldn't be too paranoid about that.

reaperrr commented:
Updated.

pchote commented:
`Exception of type `System.InvalidOperationException`: Can't find EnablePower in notification pool.` when trying to power down a structure in TS.

pchote commented:
Something seems to be wrong with the low power triggering in TD.  The following structures don't disable on low power: TMPL, OBLI, SAM, ATWR but HQ and EYE do disable.

pchote commented:
D2K and RA appear to have the same issue as TD.

reaperrr commented:
@pchote If I remove `[FieldLoader.Require]` from `ValidPowerStates` without changing anything else, everything works again.

reaperrr commented:
Updated. Fixed the TS crash and replaced `[FieldLoader.Require]` with a TODO comment for now.

pchote reviewed (dismissed):
With some more digging I found that the real problem wasn't with `ValidPowerStates` after all, but with buildings that were placed while the player was already in low power.  Adding an `Update` call to `Created` fixed the problem.

I have fixed that, removed that `// TODO:`, rebased, and updated the upgrade rule date.

The code looks good and I was not able to spot any further regressions in any of the default mods.

penev92 commented on mods/d2k/rules/structures.yaml:
```diff
@@ -629,6 +624,7 @@ starport:
 		SpawnOffset: 0,-480,0
 		ExitCell: 0,2
 	ProductionAirdrop:
+		RequiresCondition: !disabled
```
`ProductionAirdrop` is not conditionable, so this line is pointless.

penev92 reviewed (commented):
Some small nits to fix.

reaperrr commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnPowerState.cs:
```diff
@@ -0,0 +1,88 @@
+		public override object Create(ActorInitializer init) { return new GrantConditionOnPowerState(init.Self, this); }
+	}
+
+	public class GrantConditionOnPowerState : ConditionalTrait<GrantConditionOnPowerStateInfo>, INotifyOwnerChanged, INotifyPowerLevelChanged
```
The condition is provided as long as that power state is maintained, so I'd prefer to keep that name.

reaperrr committed:
Some internal renamings in power-related traits

reaperrr commented:
Updated.

penev92 reviewed (approved):
👍

pchote commented:
This needs a changelog entry!

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/76de4b13664e551388018e601f05b608dbc46216...d4ed697ea7846b8b14e66235a7154ed6f4588048
